opengovsg/postmangovsg
# Email template logos

## Problem

Continuation of #1242 

Closes [#1203, #1216]

## Solution

- Add agency logo to HTML template
- Updated `agencies` table to include `name` and `logo_uri` columns, add `domains` table to store mappings from domains to agencies
- Added backend logic to lookup agency details
- Added backend logic to populate domain and agency for new users (NOTE: does not backfill values for existing users)

## Before & After Screenshots
_Note: Copy in screenshots is outdated_

**Desktop View**

<img src="https://user-images.githubusercontent.com/9080115/121170595-016d9b80-c888-11eb-9559-3afb78489104.png" width="500"/>

**Mobile View**
<img src="https://user-images.githubusercontent.com/9080115/121160756-f19d8980-c87e-11eb-8a9d-ff344e1218f5.png" width="300"/>

**Show/Hide logo**
![image](https://user-images.githubusercontent.com/9080115/124416123-df810f00-dd88-11eb-8214-5ac3c2b17a2f.png)

**Show/Hide logo if logo is not uploaded**
![image](https://user-images.githubusercontent.com/9080115/124416068-bd878c80-dd88-11eb-8f4c-095be6ebc869.png)

## Tests

- Build `backend` code
- Run ```npx sequelize db:migrate``` from `backend`
- Email creation form should show 'Your agency’s logo has not been uploaded. To upload a logo, please fill up this form.' for agency without logo (e.g. before populating DB)

---

- Populate `domains` and `agencies` database with agency/logo details
- Email template preview in steps 2 and 4 - should show/hide logo based on option in step 1
- Send test email (step 3) - should show/hide logo based on option in step 1
- Send email campaign (step 4) - should show/hide logo based on option in step 1 

## Deploy Notes

- Motivation behind new schema is [here](https://github.com/opengovsg/postmangovsg/issues/1203#issuecomment-852788863)
```
cd backend && npx sequelize db:migrate
```
- [ ] Populate list of agencies and domains
- [ ] After populating domains, backpatch existing users table with domains (if valid)
```sql
UPDATE users 
  SET email_domain = d.domain
  FROM domains d
  WHERE SUBSTR(users.email, POSITION( '@' in users.email ) )  = d.domain;
```

## Rollback
DB migrations are backwards compatible, no rollback is required (though undoing the migrations will do no harm)

**NOTE:** Frontend MUST be rolled back before backend.
New frontend is not compatible with old backend due to the new field `showLogo` (it will fail the validator)
